collections/_posts/2022-11-14-DeformableConvolutionalNetworks.md

```diff
@@ -0,0 +1,130 @@
+* The proposed method is interesting for classification, segmentation and object tracking tasks.
+&nbsp;
+
+
+# Introduction 
+
+For a neural network, geometric transformation, viewpoint and partial deformation are difficult to learn. Most of the time, to generalise an architecture during learning, known geometric deformations such as scaling, rotation or shearing are applied. 
+However, there is no mechanism to help the model learn the geometric variations. The data augmentation is therefore limited by the user's knowledge. The proposed mechanism is end-to-end, adaptable to each architecture containing CNNs and allows CNNs to learn more complex deformations.
+
+&nbsp; 
+# Method
+
+## Basic convolution and limitation
+
+The following illustration is a simple convolution with a kernel size of $$ 3 \times 3$$ and a stride of $$1 \times 1$$. This calculation can be written as follows:
+
+$$ y(p_0) = \sum_{p_n \in R}w(p_n) \cdot x(p_0 + p_n) $$
+
+With $$p_n$$, The receptive field of the predefined convolution in the grid $$ R $$. In this specific case, the grid is equal to:
+
+$$ R = \{ (-1,-1), (-1, 0), (0, -1), ..., (0, 1), (1, 0), (1, 1) \} $$
+
+The grid $$ R $$ is static during training and inference. Thus the recepteive field of classical convolution is limited. To increase it, the kernel size, the pitch or the number of cascaded convolution can be increased.
+
+![](/collections/images/DeformableConvolutionalNetworks/standart_convolution.gif)
+
+## Deformable convolution
+
+In the proposed approach, the convolution operation can be rewritten as follows:
+
+$$ y(p_0) = \sum_{p_n \in R}w(p_n) \cdot x(p_0 + p_n + \Delta p_n) $$
+
+$$ \Delta p_n $$ correspond to the learnable grid. For $$ 3 \times 3$$ convolution, we have:
+
+$$ R = \{ (\Delta {x^1}, \Delta {y^1}), (\Delta {x^2}, \Delta {y^2})), ..., (\Delta {x^9}, \Delta {y^9}) \} $$
+
+In practice, additional layers are added to learn these components. For a $$ 3 \times 3 $$ convolution, 18 additional parameters are learned. The illustration below shows the mechanism:
+
+![](/collections/images/DeformableConvolutionalNetworks/deformable_convolution_illustration.jpg)
+
+An interesting point is that the sampling position is adaptable to the geometric transformation, especially in the sub-images below (c,d), where scaling and rotation have been applied.
+
+![](/collections/images/DeformableConvolutionalNetworks/sampling_location.jpg)
+
+In this scenario, the sample positions do not need to be integers. For fast calculation, the bilinear interpolation is implemented as follows:
+
+$$ x(p) = \sum_q max(0,1 - |p_x - q_x| ) \cdot max(0,1 - |p_y - q_y| )\cdot x(q)$$
+
+With $$ p = p_0 + p_n + \Delta p_n$$.
+&nbsp;
+
+## Deformable RoI Pooling (for Region Proposol Network)
+
+**RoI pooling**
+
+The same principle can be applied on pooling operation. Basic average RoI pooling is given by the following equation:
+
+$$ y(i,j) = \sum_{p \in bin(i, j)} \frac{ x(p_0 + p_n)}{n_{ij}} $$ 
```

Comment:
I'm not 100% sure, but since we're talking about averaging here, I think the division is meant to be _outside_ the sum, and not applied on each element of the sum. Though I must admit that the original paper is not clear on that front...

P.S. This comment also applies to the next equations.

Reply:
I am not 100% sure, but I think we can speak about RoI average pooling because the $$ n_{i,j}$$ is function of the coordonate. If this value is constant it can be out of the sum and it will produce the same result. We can discuss this later :)